src/extensions/wcf/Microsoft.DotNet.UpgradeAssistant.Extensions.WCFUpdater/UpdateRunner.cs

```diff
@@ -0,0 +1,173 @@
+            }
+        }
+
+        public static SyntaxNode? SourceCodeUpdate(SourceCodeUpdater sourceCodeUpdater, ILogger logger)
+        {
+            try
+            {
+                logger.LogDebug("Start updating source code...");
+                var root = sourceCodeUpdater.UpdateDirectives();
+                root = sourceCodeUpdater.RemoveOldCode(sourceCodeUpdater.AddTemplateCode(root));
+                logger.LogInformation("Finish updating source code.");
+                return root;
+            }
+            catch (Exception e)
+            {
+                logger.LogError(e, "Error updating source code");
+                return null;
+            }
+        }
+
+        public static void WriteSourceCodeUpdate(SyntaxNode root, string path, ILogger logger)
+        {
+            try
+            {
+                logger.LogDebug("Start writing changes to source code to update Main()...");
+                File.WriteAllText(path, root.ToFullString());
+                logger.LogInformation("Finish writing changes to source code/Main().");
+            }
+            catch (Exception e)
+            {
+                logger.LogError(e, $"Error writing changes to {path}");
+            }
+        }
+
+        public static List<SyntaxNode>? DirectiveUpdate(List<SourceCodeUpdater> list, ILogger logger)
+        {
+            try
+            {
+                logger.LogDebug("Start updating directives...");
+                var result = new List<SyntaxNode>();
+                foreach (var updater in list)
+                {
+                    result.Add(updater.UpdateDirectives());
+                }
+
+                logger.LogInformation("Finish updating directives.");
+                return result;
+            }
+            catch (Exception e)
+            {
+                logger.LogError(e, "Error updating using directives");
+                return null;
+            }
+        }
+
+        public static void WriteDirectiveUpdate(List<SyntaxNode> root, IEnumerable<string> path, ILogger logger)
+        {
+            try
+            {
+                logger.LogDebug("Start writing changes to source code to update directives...");
+                if (root.Count != path.Count())
+                {
+                    logger.LogError("The number of files does not match the number of path. Cannot write directive updates.");
+                    return;
+                }
+
+                for (int i = 0; i < root.Count; i++)
+                {
+                    File.WriteAllText(path.ElementAtOrDefault(i), root[i].ToFullString());
+                }
+
+                logger.LogInformation("Finish writing changes to source code/directives.");
+            }
+            catch (Exception e)
+            {
+                logger.LogError(e, $"Error writing changes to {path.ToList()}");
+            }
+        }
+
+        public static Dictionary<string, object> GetContext(ConfigUpdater configUpdater)
```

Comment:
Rather than use a Dictionary for this, could we have a simple type for storing WcfUpgradeContext? It would be more reliable (if something changes in that model type, things won't build until all users agree on the shape of it unlike a Dictionary which would happily build and then fail at run time).

Reply:
Hi Mike! That's a great suggestion. And I will implement it in my next PR (one with all extra features I implemented) to avoid repeated merge conflicts. I have summarized those improvements I will make in this issue: https://github.com/dotnet/upgrade-assistant/issues/1270